common/shared.c

```diff
@@ -80,47 +80,62 @@ static int getaddr(char *node, char *service, void **addr,
 		return 0;
 	}
 
-	ret = getaddrinfo(node, service, NULL, &ai);
+	ret = fi_getinfo(FT_FIVERSION, node, service, flags,
+			NULL, &fi);
 	if (ret) {
-		FT_ERR("getaddrinfo error %s\n", gai_strerror(ret));
+		FT_ERR("fi_getinfo error %s\n", fi_strerror(ret));
 		return ret;
 	}
 
-	if ((*addr = malloc(ai->ai_addrlen))) {
-		memcpy(*addr, ai->ai_addr, ai->ai_addrlen);
-		*len = (size_t)ai->ai_addrlen;
+	if (flags & FI_SOURCE) {
+		if ((*addr = malloc(fi->src_addrlen))) {
+			memcpy(*addr, fi->src_addr,
+					fi->src_addrlen);
+			*len = fi->src_addrlen;
+		} else {
+			FT_ERR("addr allocation failed\n");
+			ret = EAI_MEMORY;
+		}
```

Comment:
Is the intent that the addr and len be copied back into hints?  If so, I think you need to copy back the addr_format that the provider returns.  On a subsequent call to fi_getinfo(), a different provider than the one that returned the source address may interpret the address incorrectly.